airflow/gcp/hooks/dataflow.py

```diff
@@ -515,8 +530,20 @@ def label_formatter(labels_dict):
             return ['--labels={}={}'.format(key, value)
                     for key, value in labels_dict.items()]
 
-        self._start_dataflow(variables, name, [py_interpreter] + py_options + [dataflow],
-                             label_formatter, project_id)
+        if py_requirements is not None:
+            with TemporaryDirectory(prefix='dataflow-venv') as tmp_dir:
+                py_interpreter = prepare_virtualenv(
```

Comment:
1) would we be able to specify a pip.conf (in case of wanting to install from private pypi server)?
2) This re-instantiation / pip install every time the task runs adds potentially a lot of latency before pipeline submission (and pip install logging spam) if installing large/many packages. 
It would be good if we could re-use these virtualenvs effectively caching some dependencies (rather than re-installing from pip all the time from a bare virtualenv). However, there'd have to be a mechanism to check if this worker has a copy of the virtualenv necessary .

Reply:
1. If you want to set up a private pip repository for Cloud Composer, then you should follow these instructions:
https://cloud.google.com/composer/docs/how-to/using/installing-python-dependencies#install-private
You will also need to set the ``PIP_CONFIG_FILE`` environment variable when you use the virtual environment
2. Unfortunately, in distributed systems we should not save files on the current environment and rely on the fact that they will be available on restart. This can cause many problems, e.g. taking up all the disk space.
Instead, I recommend delivering packages from a local folder that you have previously set up. I have prepared a script that shows how to download all packages to the directory and then install them in a virtual environment
```bash
#!/bin/bash -x

TMP_PYPI="$(mktemp -d)"
TMP_VENV="$(mktemp -d)"

virtualenv ${TMP_PYPI}/venv
virtualenv ${TMP_VENV}

START_FETCH=$(date +%s)
source ${TMP_VENV}/bin/activate
mkdir -p "${TMP_PYPI}/packages"
pip download apache-beam[gcp] -d "${TMP_PYPI}/packages"
END_FETCH=$(date +%s)

START_INSTALL=$(date +%s)
source ${TMP_VENV}/bin/activate
pip install $(find "${TMP_PYPI}/packages" -type f)
END_INSTALL=$(date +%s)

echo "Fetch time: $(( END_FETCH - START_FETCH )) seconds"
echo "Install time: $(( END_INSTALL - START_INSTALL )) seconds"
```
For me, the download process took a long time, but the local installation was significantly faster.
```
Fetch time: 41 seconds
Install time: 20 seconds
```

Reply:
Interesting. For Composer, it'd be useful to put these downloaded files in the `/data/` directory on GCS which always gets synced to workers, and have the install happen from there rather than always redownloading.
https://cloud.google.com/composer/docs/concepts/cloud-storage
Agree this may be composer specific optimization and not appropriate for OSS trunk.